### ISSUE TITLE
Liquibase not checking for already deployed changesets

### ISSUE THREAD

**Issue by kailash558** [liquibase/liquibase#4946]
### Search first

- [X] I searched and no similar issues were found

### Description

I’m using Liquibase 4.21.0 to deploy changes on AWS RDS MySQL Community/Aurora MySQL clusters (MySQL 8.0).
Liquibase is deploying already executed changesets which are the same as previously executed but it's working as expected with MySQL 5.7.

The same thing is working as expected while working with Liquibase 4.8.0 in another environment with both MySQL versions.

Liquibase Command -
`PATH_LIQUIBASE_CLASSPATH = /home/ubuntu/workspace/Liquibase/mysql-connector-java-8.0.27.jar`

`liquibase --driver=com.mysql.cj.jdbc.Driver --classpath={PATH_LIQUIBASE_CLASSPATH} --url=“jdbc:mysql://{host}:3306/{database}?autoReconnect=true&useSSL=false” --username={MYSQL_USERNAME} --password={MYSQL_PASSWORD} --changelog-file={sql_file} --sql-log-level=info --logLevel=info update`

First Execution -
![First Execution with Liquibase](https://github.com/liquibase/liquibase/assets/31000726/ca6c2703-cd29-4bc9-b838-c3c6797ddf9a)

Second Execution -
![Second Execution with Liquibase](https://github.com/liquibase/liquibase/assets/31000726/f9380939-4176-4245-ba1d-9ff8ead28936)

It also creates duplicate entries in DATABASECHANGELOG table - 
![duplicaet_enteries_databasechangelog](https://github.com/liquibase/liquibase/assets/31000726/09312e27-07cd-469f-9b75-27741c9f6f5d)

### Steps To Reproduce

Liquibase command -

`PATH_LIQUIBASE_CLASSPATH = /home/ubuntu/workspace/Liquibase/mysql-connector-java-8.0.27.jar`

`liquibase --driver=com.mysql.cj.jdbc.Driver --classpath={PATH_LIQUIBASE_CLASSPATH} --url=“jdbc:mysql://{host}:3306/{database}?autoReconnect=true&useSSL=false” --username={MYSQL_USERNAME} --password={MYSQL_PASSWORD} --changelog-file={sql_file} --sql-log-level=info --logLevel=info update`

MySQL Server and Version -
AWS RDS MySQL Community/Aurora MySQL 8.0

Liquibase Script -
```
--liquibase formatted sql

--changeset kkumawat:createTable-employee_table failOnError:false_1 context:"Creating new table employee_table"
CREATE TABLE employee_table(
    id int NOT NULL AUTO_INCREMENT,
    name varchar(45) NOT NULL,
    occupation varchar(35) NOT NULL,
    age int NOT NULL,
    PRIMARY KEY (id)
);

--changeset kkumawat:alterTable-tutorials_tbl_1 failOnError:false context:"altering table tutorials_tbl, adding new column tutorial_sub_title"
alter table tutorials_tbl add column tutorial_sub_title varchar(20);

--changeset kkumawat:alterTable-employee_table_last_name_1 failOnError:false context:"altering table employee_table, adding new column last_name"
alter table employee_table add column last_name varchar(20);

--changeset kkumawat:insertIntoTable-employee_table_1 failOnError:false context:"inserting in to table employee_table"
insert into employee_table (name, occupation, age) values ("Kailash", "Software Engineer", 25);

--changeset kkumawat:selectFromTable-employee_table_1 failOnError:false context:"select and store data form table employee_table"
select age into @value1 from employee_table where name="Kailash";

--changeset kkumawat:selectFromTable-employee_table_value1_1 failOnError:false context:"select and store data form table employee_table"
select @value1;

--changeset kkumawat:insertIntoTable-employee_table_value1_1 failOnError:false context:"inserting in to table employee_table"
insert into employee_table (name, occupation, age) values ("Kailash K", "Software Engineer", @value1);
```

### Expected/Desired Behavior

Liquibase should not deploy already executed changesets unless runAlways is set to True.

### Liquibase Version

4.21.0

### Database Vendor & Version

AWS RDS MySQL Community/Aurora MySQL 8.0

### OS and/or Infrastructure Type/Provider

EC2

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR (Thank you!)

**Comment by tati-qalified** [comment]
Hello @kailash558. I've tested this issue with AWS Aurora and with MySQL in a docker container, but I haven't been able to reproduce the bug. In my computer, running update for a second time doesn't run any changesets.

Would you be able to re-test it, and check that there isn't another thing that's causing the problem? Perhaps something in your AWS instance? Let me know the results and we can go from there.

Thank you
Tatiana

**Comment by tati-qalified** [comment]
Hello @kailash558, I've tested this issue with the code you've provided, Liquibase v4.21.0 and MySQL v8.0, and I haven't been able to replicate the bug. 
Looking into the logs, I can see that the deployment id is different for both instances. According to Liquibase documentation, the deployment id in the databasechangelog table signifies `changesets deployed together will have the same unique identifier.`
As they are different, it may be something relating to the path where the files are. Is there any chance that the path could be changing in between runs? Is any part of it dynamic?
Please re-test the issue and let me know what you find.
Thank you
Tatiana

**Comment by kailash558** [comment]
Hi @tati-qalified, 
There is no dynamic file creation. All changeset files are prepared manually and placed in Gitlab. Then program automatically picks them to deploy. 
I'm also not sure how can it be reproduced in some other system.

**Comment by tati-qalified** [comment]
Alright. I'll need some more info from you, and I'll explain the steps I took to test this so you can let me know if you're doing anything differently.

**Questions:**
- Is GitLab running `liquibase update`, or do you have a root changelog that links to GitLab?
- Could you be missing any permissions for the database that would make Liquibase get the `databasechangelog` table as empty?

**Steps taken to test**
- Create a changelog with your sample code
- Create a `liquibase.properties` file: here I referenced the changelog file directly in the same folder
- Run `liquibase update`
- Output:
```
UPDATE SUMMARY
Run:                          7
Previously run:               0
Filtered out:                 0
-------------------------------
Total change sets:            7
```
- Run `liquibase update` again
- Output:
```
UPDATE SUMMARY
Run:                          0
Previously run:               7
Filtered out:                 0
-------------------------------
Total change sets:            7
```

You can see that the changesets aren't being executed twice.

Please re-test this issue and let me know if there's anything that you're doing differently.
Thanks,
Tatiana

**Comment by kailash558** [comment]
Hi Tatiana,

**Answers to Questions -**

- Is GitLab running liquibase update, or do you have a root changelog that links to GitLab?
GitLab is not running Liquibase update. We add manually prepared .sql files (Liquibase formatted SQL files) in GitLab. Our python program picks those SQL files and then deploys them. Liquibase update command is run by python script using subprocess module. 

- Could you be missing any permissions for the database that would make Liquibase get the databasechangelog table as empty?
There is no issue with permissions. It has all required permissions. Liquibase is able to insert changelog data, read information about change log lock.


**Steps taken to test -**
1. Add manually prepared Liquibase formatted SQL files in Gitlab
2. Run Python program from Jenkins job. (We have added Liquibase path in $PATH variable)
3. Python program pull SQL files from Gitlab
4. Script runs liquibase update command through subprocess module and get result.

We just run same Jenkins Job without any change in SQL file or script or Jenkins configuration.

Thanks, 
Kailash

**Comment by kevin-atx** [comment]
@tati-qalified - I moved this ticket back to you for review. This looks similar to the problem related to the DATABASECHANGELOG uppercase to lowercase change you see with BigQuery in ticket #5034

**Comment by filipelautert** [comment]
@tati-qalified  @kailash558 I saw similar issues in the past and they were related to PATH not matching when they start/contains "." or ".." . This should be fixed on 4.25.0

**Comment by tati-qalified** [comment]
With Filipe's comment I was finally able to reproduce the issue. It seems like this only happens when:

- Using MySQL 8.0
- The main changelog is pointing to this executed changelog, which is one directory above the main one
![image](https://github.com/liquibase/liquibase/assets/135264316/5c6f0f27-c09a-4e21-b791-6462e8b18a38)
- The changesets have the attribute `failOnError` set to `false`

If any of these things aren't happening, everything works as expected.

As Filipe said, this is a problem related to the path that contains "./../", which was a known issue. As `failOnError` was set to `false`, the execution continued even after the changes failed, and that's why there were duplicate entries in the `databasechangelog` table.

This issue has been fixed as of Liquibase 4.25.0.

In summary, @kailash558 I'd recommend you to update to the latest Liquibase version, re-test this and let us know if it works for you.

Thank you for reporting this issue!
Tatiana
